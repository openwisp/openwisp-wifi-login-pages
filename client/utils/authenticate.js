import {sessionStorage} from "./storage";

const authenticate = (cookies, orgSlug) => {
  const token = cookies.get(`${orgSlug}_auth_token`);
  const sessionKey = sessionStorage.getItem(`${orgSlug}_auth_token`);
  if (sessionKey) {
    cookies.remove(`${orgSlug}_auth_token`, {path: "/"});
    cookies.remove(`${orgSlug}_username`, {path: "/"});
    return true;
  }
  if (token) return true;
  return false;
};
export default authenticate;
