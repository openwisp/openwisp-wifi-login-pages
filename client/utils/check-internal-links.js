const isInternalLink = (link) => {
  if (link.length && link[0] === "/") return true;
  return false;
};
export default isInternalLink;
