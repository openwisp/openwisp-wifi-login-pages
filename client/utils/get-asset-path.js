const getAssetPath = (slug, relativePath) => `/assets/${slug}/${relativePath}`;
export default getAssetPath;
