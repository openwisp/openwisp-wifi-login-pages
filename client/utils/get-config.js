const loadConfig = (slug) => import(`../configs/${slug}.json`);
const getConfig = async (slug) => {
  let config;
  try {
    config = await loadConfig(slug);
  } catch (err) {
    console.error(`Configuration for organization ${slug} does not exists.`);
    config = undefined;
  }
  return config;
};

export default getConfig;
