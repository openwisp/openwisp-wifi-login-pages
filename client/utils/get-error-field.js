import React from "react";

const getErrorField = (errors, key, nonField = false, className = "") =>
  errors[key] ? (
    <div className={!nonField ? `error ${className}` : "error non-field"}>
      <span className="icon">!</span>
      <span className="text">{errors[key]}</span>
    </div>
  ) : null;

export default getErrorField;
