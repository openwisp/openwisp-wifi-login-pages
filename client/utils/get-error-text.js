const getErrorText = (error, fallbackVal = "") => {
  const {data} = error.response;
  // eslint-disable-next-line no-nested-ternary
  return data.detail
    ? data.detail
    : data.non_field_errors
    ? data.non_field_errors[0]
    : fallbackVal;
};
export default getErrorText;
