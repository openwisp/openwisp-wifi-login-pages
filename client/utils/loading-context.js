import React from "react";

export const loadingContextValue = {setLoading: () => {}, getLoading: () => {}};
const LoadingContext = React.createContext(loadingContextValue);

export default LoadingContext;
