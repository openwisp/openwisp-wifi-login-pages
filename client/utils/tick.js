const tick = function () {
  return new Promise((resolve) => {
    process.nextTick(resolve);
  });
};

export default tick;
