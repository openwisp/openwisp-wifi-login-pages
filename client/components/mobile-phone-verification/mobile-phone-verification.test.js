/* eslint-disable prefer-promise-reject-errors */
/* eslint-disable camelcase */
import axios from "axios";
import {shallow} from "enzyme";
import {toast} from "react-toastify";
import React from "react";
import PropTypes from "prop-types";
import {Cookies} from "react-cookie";
import {loadingContextValue} from "../../utils/loading-context";
import tick from "../../utils/tick";
import getConfig from "../../utils/get-config";
import MobilePhoneVerification from "./mobile-phone-verification";
import validateToken from "../../utils/validate-token";

jest.mock("../../utils/get-config");
jest.mock("../../utils/validate-token");
jest.mock("axios");

const createTestProps = function (props, configName = "test-org-2") {
  const config = getConfig(configName);
  return {
    mobile_phone_verification: config.components.mobile_phone_verification_form,
    settings: config.settings,
    orgSlug: config.slug,
    language: "en",
    cookies: new Cookies(),
    logout: jest.fn(),
    setUserData: jest.fn(),
    userData: {},
    ...props,
  };
};

const userData = {
  response_code: "AUTH_TOKEN_VALIDATION_SUCCESSFUL",
  radius_user_token: "o6AQLY0aQjD3yuihRKLknTn8krcQwuy2Av6MCsFB",
  username: "tester@tester.com",
  is_active: false,
  is_verified: false,
  phone_number: "+393660011222",
};

const createShallowComponent = function (props) {
  return shallow(<MobilePhoneVerification {...props} />, {
    context: {...loadingContextValue},
  });
};

describe("Mobile Phone Token verification: standard flow", () => {
  let props;
  let wrapper;
  let lastConsoleOutuput;
  let originalError;
  const event = {preventDefault: jest.fn()};

  beforeEach(() => {
    MobilePhoneVerification.contextTypes = {
      setLoading: PropTypes.func,
    };
    props = createTestProps();
    axios.mockImplementationOnce(() => {
      return Promise.resolve({
        status: 201,
        statusText: "CREATED",
        data: null,
      });
    });
    // console mocking
    validateToken.mockClear();
    originalError = console.error;
    lastConsoleOutuput = null;
    console.error = (data) => {
      lastConsoleOutuput = data;
    };
  });

  afterEach(() => {
    axios.mockReset();
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
    sessionStorage.clear();
    console.error = originalError;
  });

  it("should render successfully", async () => {
    validateToken.mockReturnValue(true);
    jest.spyOn(MobilePhoneVerification.prototype, "createPhoneToken");

    wrapper = createShallowComponent(props);
    wrapper.setProps({userData});
    await tick();

    expect(axios).toHaveBeenCalled();
    expect(
      MobilePhoneVerification.prototype.createPhoneToken,
    ).toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
    const css = "form";
    expect(wrapper.find(css)).toHaveLength(1);
    expect(wrapper.find(`${css} button[type='submit']`)).toHaveLength(1);
    expect(wrapper.find(`${css} input[type='text']`)).toHaveLength(1);
    expect(
      wrapper.find("form .row .label").text().includes("+393660011222"),
    ).toBe(true);
    expect(wrapper.find(".resend .button")).toHaveLength(1);
    expect(wrapper.find(".change .button")).toHaveLength(1);
    expect(wrapper.find(".logout .button")).toHaveLength(1);
    expect(wrapper.instance().hasPhoneTokenBeenSent()).toBe(true);
  });

  it("should resend token successfully", async () => {
    jest.spyOn(MobilePhoneVerification.prototype, "resendPhoneToken");
    jest.spyOn(toast, "info");
    validateToken.mockReturnValue(true);
    wrapper = createShallowComponent(props);
    await tick();

    axios.mockImplementationOnce(() => {
      return Promise.resolve({
        status: 201,
        statusText: "CREATED",
        data: null,
      });
    });
    wrapper.find(".resend .button").simulate("click");
    expect(
      MobilePhoneVerification.prototype.resendPhoneToken.mock.calls.length,
    ).toBe(1);
    expect(toast.info.mock.calls.length).toBe(1);
  });

  it("should verify token successfully and must call setUserData", async () => {
    jest.spyOn(MobilePhoneVerification.prototype, "handleSubmit");
    validateToken.mockReturnValue(true);
    wrapper = createShallowComponent(props);
    wrapper.setProps({userData});
    const setUserDataMock = wrapper.instance().props.setUserData.mock;
    await tick();

    axios.mockImplementationOnce(() => {
      return Promise.resolve({
        status: 200,
        statusText: "OK",
        data: null,
      });
    });
    wrapper
      .find("form .code input[type='text']")
      .simulate("change", {target: {value: "12345", name: "code"}});
    expect(wrapper.instance().state.code).toBe("12345");
    wrapper.find("form").simulate("submit", event);
    await tick();
    expect(setUserDataMock.calls.length).toBe(1);
    expect(setUserDataMock.calls.pop()).toEqual([
      {
        ...userData,
        is_active: true,
        is_verified: true,
        justAuthenticated: true,
      },
    ]);
    expect(
      MobilePhoneVerification.prototype.handleSubmit.mock.calls.length,
    ).toBe(1);
    expect(event.preventDefault).toHaveBeenCalled();
  });

  it("should show errors", async () => {
    jest.spyOn(MobilePhoneVerification.prototype, "handleSubmit");
    validateToken.mockReturnValue(true);
    wrapper = createShallowComponent(props);
    const setUserDataMock = wrapper.instance().props.setUserData.mock;
    await tick();
    expect(setUserDataMock.calls.length).toBe(0);
    axios.mockImplementationOnce(() => {
      return Promise.reject({
        response: {
          status: 400,
          statusText: "BAD REQUEST",
          data: {
            non_field_errors: ["Invalid code."],
          },
        },
      });
    });
    wrapper
      .find("form .code input[type='text']")
      .simulate("change", {target: {value: "12345", name: "code"}});
    expect(wrapper.instance().state.code).toBe("12345");
    wrapper.find("form").simulate("submit", event);
    await tick();
    expect(setUserDataMock.calls.length).toBe(0);
    expect(
      MobilePhoneVerification.prototype.handleSubmit.mock.calls.length,
    ).toBe(1);
    expect(event.preventDefault).toHaveBeenCalled();
    expect(wrapper.instance().state.errors.nonField).toBeTruthy();
    expect(lastConsoleOutuput).not.toBe(null);
  });

  it("should log out successfully", async () => {
    jest.spyOn(MobilePhoneVerification.prototype, "handleLogout");
    validateToken.mockReturnValue(true);
    jest.spyOn(toast, "success");

    wrapper = createShallowComponent(props);
    await tick();

    wrapper.find(".logout .button").simulate("click");
    await tick();
    expect(
      MobilePhoneVerification.prototype.handleLogout.mock.calls.length,
    ).toBe(1);
    expect(wrapper.instance().props.logout.mock.calls.length).toBe(1);
    expect(toast.success.mock.calls.length).toBe(1);
  });
});

describe("Mobile Phone Token verification: corner cases", () => {
  let props;
  let wrapper;
  beforeEach(() => {
    MobilePhoneVerification.contextTypes = {
      setLoading: PropTypes.func,
    };
    props = createTestProps();
    validateToken.mockClear();
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
    sessionStorage.clear();
  });

  it("should not proceed if user is already verified", async () => {
    jest.spyOn(MobilePhoneVerification.prototype, "createPhoneToken");
    validateToken.mockReturnValue(true);
    wrapper = createShallowComponent(props);
    wrapper.setProps({
      userData: {...userData, is_active: true, is_verified: true},
    });
    await tick();
    expect(
      MobilePhoneVerification.prototype.createPhoneToken,
    ).not.toHaveBeenCalled();
    expect(wrapper.instance().state.phone_number).toBe("+393660011222");
  });

  it("should not proceed if mobile verification is not enabled", async () => {
    jest.spyOn(MobilePhoneVerification.prototype, "createPhoneToken");
    validateToken.mockReturnValue(true);
    props.settings.mobile_phone_verification = false;
    wrapper = createShallowComponent(props);
    wrapper.setProps({props});
    await tick();
    expect(
      MobilePhoneVerification.prototype.createPhoneToken,
    ).not.toHaveBeenCalled();
  });
});
