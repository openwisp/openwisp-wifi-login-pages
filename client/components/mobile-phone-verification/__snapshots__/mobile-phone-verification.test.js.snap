// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MobilePhoneVerification /> rendering with placeholder translation tags should render translation placeholder correctly 1`] = `
<div
  className="container content"
  id="mobile-phone-verification"
>
  <div
    className="inner"
  >
    <div
      className="main-column"
    >
      <div
        className="inner"
      >
        <form
          className=""
          onSubmit={[Function]}
        >
          <div
            className="row fieldset code"
          >
            <p
              className="label"
            >
              PHONE_VERIFY ()
            </p>
            <div
              className="row"
            >
              <input
                className="input "
                id="code"
                name="code"
                onChange={[Function]}
                pattern="[0-9]{6,}"
                placeholder="MOBILE_CODE_PHOLD"
                required={true}
                title="MOBILE_CODE_TITL"
                type="text"
                value=""
              />
            </div>
            <button
              className="button full"
              type="submit"
            >
              MOBILE_PHONE_VERIFY
            </button>
          </div>
        </form>
        <div
          className="row fieldset resend"
        >
          <p
            className="label"
          >
            RESEND_TOKEN_LBL
          </p>
          <button
            className="button full"
            onClick={[Function]}
            type="button"
          >
            RESEND_TOKEN
          </button>
        </div>
        <div
          className="row fieldset change"
        >
          <p
            className="label"
          >
            PHONE_CHANGE_LBL
          </p>
          <a
            className="button full"
            href="/test-org-2/change-phone-number"
          >
            PHONE_CHANGE_BTN
          </a>
        </div>
        <div
          className="row fieldset logout"
        >
          <p
            className="label"
          >
            LOGOUT_LBL
          </p>
          <button
            className="button full"
            onClick={[Function]}
            type="button"
          >
            LOGOUT
          </button>
        </div>
      </div>
    </div>
    <Memo(Connect(Contact)) />
  </div>
</div>
`;

exports[`Mobile Phone Token verification: standard flow should render successfully 1`] = `
<div
  className="container content"
  id="mobile-phone-verification"
>
  <div
    className="inner"
  >
    <div
      className="main-column"
    >
      <div
        className="inner"
      >
        <form
          className=""
          onSubmit={[Function]}
        >
          <div
            className="row fieldset code"
          >
            <p
              className="label"
            >
              Verify your mobile phone number (+393660011222) by inserting the code you received via SMS below.
            </p>
            <div
              className="row"
            >
              <input
                className="input "
                id="code"
                name="code"
                onChange={[Function]}
                pattern="[0-9]{6,}"
                placeholder="enter SMS verification code"
                required={true}
                title="this field must be a 6 digit numeric code"
                type="text"
                value=""
              />
            </div>
            <button
              className="button full"
              type="submit"
            >
              Submit
            </button>
          </div>
        </form>
        <div
          className="row fieldset resend"
        >
          <p
            className="label"
          >
            If you did not receive the verification SMS code, you can try sending it again by clicking below.
          </p>
          <button
            className="button full"
            onClick={[Function]}
            type="button"
          >
            Send a new verification code
          </button>
        </div>
        <div
          className="row fieldset change"
        >
          <p
            className="label"
          >
            Wrong phone number?
          </p>
          <a
            className="button full"
            href="/test-org-2/change-phone-number"
          >
            Change your phone number
          </a>
        </div>
        <div
          className="row fieldset logout"
        >
          <p
            className="label"
          >
            If you need to cancel the process you can log out below.
          </p>
          <button
            className="button full"
            onClick={[Function]}
            type="button"
          >
            Logout
          </button>
        </div>
      </div>
    </div>
    <Connect(Contact) />
  </div>
</div>
`;
